### PR TITLE
Epoch tree arity

### LINKED ISSUE #138: Increase arity of epoch tree
Increasing the arity of the epoch tree should significantly reduce the proving time for most of our circuits. We'll need custom contracts, circuits and probably changes in the core package.

### PR DESCRIPTION
**Pull request by vimwitch**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`yarn build`) was run locally and any changes were pushed
- [x] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Closes #138

## What is the new behavior?

Supports variable arity for the epoch tree.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/utils/src/SparseMerkleTree.ts [hunk 1]
```diff
@@ -109,27 +126,30 @@ export class SparseMerkleTree {
...
+    public createProof(leafKey: bigint): bigint[][] {
         assert(
             leafKey < this.numLeaves,
             `leaf key ${leafKey} exceeds total number of leaves ${this.numLeaves}`
         )
 
-        const siblingNodeHashes: bigint[] = []
+        const siblingNodeHashes: bigint[][] = Array(this.height)
+            .fill(null)
+            .map(() => [])
```
Comment (vivianjeng) [inline packages/utils/src/SparseMerkleTree.ts:137]: I think it is the same as `Array(this.height).fill([])`

**Review by vivianjeng — APPROVED** [review]
LGTM
And also good solution for increasing the arity

File: packages/utils/src/SparseMerkleTree.ts [hunk 2]
```diff
@@ -16,17 +17,26 @@ export class SparseMerkleTree {
...
-    constructor(private _height: number, zeroHash: bigint = BigInt(0)) {
+    constructor(
+        private _height: number,
+        zeroHash: bigint = BigInt(0),
+        arity: number = 2
+    ) {
         assert(_height > 0, 'SMT height needs to be > 0')
+        assert(
+            arity >= 2 && arity <= 16,
+            'arity must be between 2 and 16 inclusive'
```
[the remark this answers is not in the document]
Comment (vimwitch) [reply packages/utils/src/SparseMerkleTree.ts:28]: It's the limit of the hashing library.

File: packages/utils/src/SparseMerkleTree.ts (hunk 1 above)
Comment (vimwitch) [reply packages/utils/src/SparseMerkleTree.ts:137]: It's not, `Array(this.height).fill([])` uses a single element by reference. It's like doing this:

```js
const a = []
const arr = [a, a, a, a]
```

If any of them get mutated they all get mutated. Need to map into a function that creates a new array for each index.

**Pull request merged**